storm-client/src/jvm/org/apache/storm/executor/spout/SpoutOutputCollectorImpl.java

```diff
@@ -105,19 +134,16 @@ public void reportError(Throwable error) {
                 msgId = MessageId.makeUnanchored();
             }
 
-            TupleImpl tuple = new TupleImpl(executor.getWorkerTopologyContext(), values, this.taskId, stream, msgId);
-            executor.getExecutorTransfer().transfer(t, tuple);
+            final TupleImpl tuple = new TupleImpl(executor.getWorkerTopologyContext(), values, executor.getComponentId(), this.taskId, stream, msgId);
+            AddressedTuple adrTuple = new AddressedTuple(t, tuple);
+            executor.getExecutorTransfer().tryTransfer(adrTuple, executor.getPendingEmits());
         }
         if (isEventLoggers) {
-            executor.sendToEventLogger(executor, taskData, values, executor.getComponentId(), messageId, random);
+            taskData.sendToEventLogger(executor, values, executor.getComponentId(), messageId, random, executor.getPendingEmits());
         }
```

Comment:
@HeartSaVioR you pointed out some optimizations are possible to this .. that we can tackle in another jira ... can you elaborate or capture your thoughts into a jira ?